src/ucp/stream/stream_recv.c

```diff
@@ -8,132 +8,399 @@
+                                            -sizeof(ucp_stream_am_hdr_t)) - 1)
+
+
+static UCS_F_ALWAYS_INLINE ucs_status_ptr_t
+ucp_stream_recv_data_nb_internal(ucp_ep_ext_stream_t *ep_stream, size_t *length)
+{
+    ucp_recv_desc_t     *rdesc;
+
+    if (ucs_unlikely(ucs_queue_is_empty(&ep_stream->data))) {
+        return UCS_STATUS_PTR(UCS_OK);
     }
 
-    rdesc = ucs_queue_pull_elem_non_empty(&ep->ext.stream->data, ucp_recv_desc_t,
+    rdesc = ucs_queue_pull_elem_non_empty(&ep_stream->data, ucp_recv_desc_t,
                                           stream_queue);
+    ucs_trace_data("dequeued %zu stream bytes", rdesc->length);
     *length = rdesc->length;
-    ret = ucp_stream_rdesc_data(rdesc);
 
-out:
+    return ucp_stream_rdesc_data(rdesc);
+}
+
+UCS_PROFILE_FUNC(ucs_status_ptr_t, ucp_stream_recv_data_nb,
+                 (ep, length), ucp_ep_h ep, size_t *length)
+{
+    ucs_status_ptr_t ret;
+
+    UCP_THREAD_CS_ENTER_CONDITIONAL(&ep->worker->mt_lock);
+    ret = ucp_stream_recv_data_nb_internal(ep->ext.stream, length);
     UCP_THREAD_CS_EXIT_CONDITIONAL(&ep->worker->mt_lock);
 
     return ret;
 }
 
+static UCS_F_ALWAYS_INLINE void
+ucp_stream_rdesc_release(ucp_recv_desc_t *rdesc)
+{
+    if (ucs_unlikely(rdesc->flags & UCP_RECV_DESC_FLAG_UCT_DESC)) {
+        uct_iface_release_desc((char*)rdesc - sizeof(ucp_eager_sync_hdr_t));
+    } else {
+        ucs_mpool_put_inline(rdesc);
+    }
+}
+
 UCS_PROFILE_FUNC_VOID(ucp_stream_data_release, (ep, data),
                       ucp_ep_h ep, void *data)
 {
     ucp_recv_desc_t *rdesc = ucp_stream_rdesc_from_data(data);
 
     UCP_THREAD_CS_ENTER_CONDITIONAL(&ep->worker->mt_lock);
 
-    if (ucs_unlikely(rdesc->flags & UCP_RECV_DESC_FLAG_UCT_DESC)) {
-        uct_iface_release_desc((char*)rdesc - sizeof(ucp_eager_sync_hdr_t));
-    } else {
-        ucs_mpool_put_inline(rdesc);
+    ucp_stream_rdesc_release(rdesc);
+
+    UCP_THREAD_CS_EXIT_CONDITIONAL(&ep->worker->mt_lock);
+}
+
+static UCS_F_ALWAYS_INLINE ssize_t
+ucp_stream_data_unpack(void *data, size_t length, ucp_request_t *req)
+{
+    /* Truncated error is not actual for stream, need to adjust */
+    size_t       valid_len = ucs_min(req->recv.length - req->recv.state.offset,
+                                     length);
+    ucs_status_t status    = ucp_dt_unpack(req->recv.datatype, req->recv.buffer,
+                                           req->recv.length, &req->recv.state,
+                                           data, valid_len,
+                                           UCP_RECV_DESC_FLAG_LAST);
+    if (ucs_likely(status == UCS_OK)) {
+        req->recv.state.offset += valid_len;
+        ucs_trace_data("unpacked %zd\n", valid_len);
+
+        return valid_len;
+    }
+
+    ucs_assert(status != UCS_ERR_MESSAGE_TRUNCATED);
+
+    return status;
+}
+
+static UCS_F_ALWAYS_INLINE ucs_status_ptr_t
+ucp_stream_request_recv_get(ucp_worker_t *worker, void *buffer, size_t *count,
```

Comment:
Minor: `ucp_stream_recv_request_get` looks better to me.
Alswo why do you pass counter by address? It is not supposed to be changed here